pandas/core/groupby/groupby.py

```diff
@@ -4740,7 +4740,10 @@ def _wrap_transformed_output(self, output, names=None):
 
     def _wrap_agged_blocks(self, items, blocks):
         if not self.as_index:
-            index = np.arange(blocks[0].values.shape[1])
+            if blocks[0].values.ndim > 1:
```

Comment:
In a case such as:
```
pd.DataFrame({'time': [pd.Timestamp('2012-01-01 13:00:00+00:00')],
              'A': [3]}).groupby('A', as_index=False).first()
```
the ``blocks[0].values`` is a ``DatetimeIndex`` and not an ``array`` so trying to call ``shape[1]`` on a ``DTI`` results in an out of index error and then the compat routine for ``first`` or ``last`` are called which leads to the timezone being lost.

Reply:
> the compat routine for first or last are called which leads to the timezone being lost.

Is there a case where DatetimeTZ data can go through the compat routine (maybe with `as_index=True`?) It would be great for the data type to also be preserved there as well.

Reply:
Yeah I have also been wondering about the same thing. The following case goes through the compat routine and consequently loses the timezone:
```
In [2]: df = pd.DataFrame({'group': [1, 1, 2],
                           'category_string': pd.Series(list('abc')).astype('category'),
                           'datetimetz': pd.date_range('20130101', periods=3, tz='US/Eastern'})
In [3]: df.groupby('group').first()
Out[3]: 
      category_string          datetimetz                                                
group                                                                                                                                                     
1                   a 2013-01-01 05:00:00                                                                      
2                   c 2013-01-03 05:00:00    
```

But if we exclude the categorical column it doesn't go through the compat routine and preserves the timezone information:
```
In[4]: df[['group', 'datetimetz']].groupby('group').first()
Out[4]:
                     datetimetz                                                       
group                                                                  
1     2013-01-01 00:00:00-05:00                                                            
2     2013-01-03 00:00:00-05:00
```

So if we have the categorical column do we want to legitimately go through the compat routine? And, if so, should we preserve the timezone in the compat routine? I think this might actually be quite straightforward (see #15885)